src/python/zquantum/qaoa/recursive_qaoa.py

```diff
@@ -0,0 +1,247 @@
+        solve the reduced QAOA problem, mapping the reduced solution to the original solution.
+
+        Args:
+            n_c: The threshold number of qubits at which recursion stops, as described in the original paper.
+                Cannot be greater than number of qubits.
+            ansatz: an Ansatz object with all params (ex. `n_layers`) initialized
+            initial_params: initial parameters used for optimization
+            optimizer: Optimizer object used for optimizer
+            estimation_tasks_factory_generator: function that generates EstimationTasksFactory objects
+                from operator and ansatz. See example below for clarification.
+            cost_function_factory: function that generates CostFunction objects given EstimationTasksFactory.
+                See example below for clarification.
+
+        Example usage (aka, what the heck are all these factories?):
+
+            from functools import partial
+            from zquantum.core.estimation import (
+                estimate_expectation_values_by_averaging,
+                allocate_shots_uniformly
+            )
+            from zquantum.core.cost_function import (
+                substitution_based_estimation_tasks_factory,
+                create_cost_function,
+            )
+
+            cost_hamiltonian = ...
+            ansatz = ...
+
+            estimation_preprocessors = [partial(allocate_shots_uniformly, number_of_shots=1000)]
+            estimation_tasks_factory_generator = partial(
+                substitution_based_estimation_tasks_factory,
+                estimation_preprocessors=estimation_preprocessors
+            )
+            cost_function_factory = partial(
+                create_cost_function,
+                backend=QuantumBackend,
+                estimation_method=estimate_expectation_values_by_averaging,
+                parameter_preprocessors=None,
+            )
+
+            initial_params = np.array([0.42, 4.2])
+            optimizer = ...
+
+            recursive_qaoa = RecursiveQAOA(
+                3,
+                ansatz,
+                initial_params,
+                optimizer,
+                estimation_tasks_factory,
+                cost_function_factory,
+            )
+
+            solutions = RecursiveQAOA(cost_hamiltonian)
+        """
+
+        self._n_c = n_c
+        self._ansatz = ansatz
+        self._initial_params = initial_params
+        self._optimizer = optimizer
+        self._estimation_tasks_factory = estimation_tasks_factory
+        self._cost_function_factory = cost_function_factory
+
+    def __call__(
+        self,
+        cost_hamiltonian: IsingOperator,
+        qubit_map: Optional[List[List[int]]] = None,
+    ) -> List[Tuple]:
+        """Args:
+            cost_hamiltonian: Hamiltonian representing the cost function.
+            qubit_map: A list that maps qubits in reduced Hamiltonian back to original qubits, used for
```

Comment:
Why is it a list and not a dictionary?

Reply:
Is there an easy and reliable way to get the nth element of a dictionary?

Reply:
I agree that hash table makes more sense in this scenario, but it is a little more complicated to use. Perhaps we could have two hash tables:  original qubit -> new qubit, and new qubit -> original qubit, while recording the sign information. Since one can construct one hash table from the other, only one of them is needed as input.

Reply:
It seems that changing qubit_map leads to a lot of extra work. Since the current approach works well, I suggest to keep it as it is.

Reply:
Haha it's ok it didn't take that long, it took about an hour, so I ended up changing it to a dictionary/hash table.